### PR TITLE
Reduce compile time/size for scan.cu

### PR DESCRIPTION
**Pull request by davidwendt**
This PR reduces the number of calls to `inclusive_scan` and `exclusive_scan` by using a `null_replace_accessor` that allows non-nullable columns. This reduces the compile time and size of `scan.cu` by half. This PR also includes a scan gbenchmark that shows no change in performance from the original implementation.

### CODE REVIEW

**Comment by davidwendt** [comment]
|  | baseline | refactored  |
| --- | --- | --- |
| compile of scan.cu | 194 s | 113 s |
| scan.cu.o size | 6.42MB | 3.54MB |
| libcudf.so size | 353MB | 351MB |
| debug compile scan.cu | 27.3 min | 13.6 min |
| debug scan.cu.o size |  361MB | 188MB |

File: cpp/src/reductions/scan.cu [hunk 1]
```diff
@@ -34,6 +33,32 @@
...
 namespace detail {
+
+namespace {
+/**
+ * @brief Accessor handles both nullable and non-nullable columns.
+ *
+ * @tparam Element type used for null-replacement value
+ */
+template <typename Element>
+struct null_replace_accessor {
```
Comment (jrhemstad) [inline cpp/src/reductions/scan.cu:44]: We already have a null replacement iterator. How is this different?

Comment (davidwendt) [reply cpp/src/reductions/scan.cu:44]: This one allows the column to not have a null-mask.
I think it should go into the `interator.cuh` but then I need to think of a name for it.

Comment (davidwendt) [reply cpp/src/reductions/scan.cu:44]: Or perhaps I can change the current one to this one with appropriate defaults.

**Comment by hyperbolic2346** [comment]
This improvement is great and much needed. Do you know how this impacts the run-time performance of scan?

**Comment by davidwendt** [comment]
For nullable columns the performance does not change -- the code is almost identical so that is expected. The main change is for non-nullable columns which did not require a null-replacement-iterator. We are adding an extra non-divergent runtime if-check when retrieving the element. I reran my benchmarks and found this approach does slow a bit with smaller rows (10K) but never more than 1-2 microseconds. For larger rows (1M or more) the slow down is in nanoseconds and less than 1.25%.

Here are some benchmark results for some non-null fixed-width columns. I used the gbenchmark included in this PR on my desktop which is Ubuntu 18.04 with CUDA 11.0 on a GV100.

| type | size | us diff | % change |
| --- | ---| ---| ---|
| int32 | 10K | 1.12 | 3.76% |
| int32 | 100K | 0.743 | 2.36% |
| int32 | 1M | 0.498 | 1.21% |
| int32 | 10M | 0.692 | 0.42% |
| int32 | 100M | -0.482 | -0.03% |
| uint64 | 10K | 0.825 | 2.72% |
| uint64 | 100K | 0.470 | 1.44% |
| uint64 | 1M | 0.362 | 0.62% |
| uint64 | 10M | 0.359 | 0.12% |
| uint64 | 100M | -0.208 | -0.01% |
| float | 10K | 1.061 | 3.55% |
| float | 100K | 0.533 | 1.68% |
| float | 1M | 0.269 | 0.65% |
| float | 10M | 0.701 | 0.43% |
| float | 100M | -0.102 | -0.01% |

The negative values means that the null-replace implementation in this PR was technically faster though within a reasonable measurement fluctuation error. All of the smaller row columns run in about 32us. Only the 100M row benchmarks are above 1ms.

Honestly, I was hoping it would be faster since we are generating half the code. But thrust's scan is very fast with fixed-width types and trivial predicates like the ones in scan.cu.

**Review by trxcllnt — APPROVED** [review]

**Review by jrhemstad — APPROVED** [review]

**Comment by kkraus14** [comment]
@gpucibot merge

**Pull request merged**